crates/red_knot_module_resolver/src/resolver.rs

```diff
@@ -152,57 +161,308 @@ impl RawModuleResolutionSettings {
+            .unwrap(),
         );
 
+        let stdlib = Arc::new(custom_typeshed.map_or_else(
+            ModuleResolutionPathBuf::vendored_stdlib,
+            |custom| {
+                ModuleResolutionPathBuf::stdlib_from_custom_typeshed_root(&SystemPath::absolute(
+                    custom,
+                    current_directory,
+                ))
+                .unwrap()
+            },
+        ));
+
         // TODO vendor typeshed's third-party stubs as well as the stdlib and fallback to them as a final step
-        if let Some(site_packages) = site_packages {
-            paths.push(ModuleResolutionPathBuf::site_packages(site_packages).unwrap());
-        }
+        let site_packages = site_packages.map(|path| {
+            Arc::new(
+                ModuleResolutionPathBuf::site_packages(SystemPath::absolute(
+                    path,
+                    current_directory,
+                ))
+                .unwrap(),
+            )
+        });
 
         ModuleResolutionSettings {
             target_version,
-            search_paths: OrderedSearchPaths(paths.into_iter().map(Arc::new).collect()),
+            search_path_settings: ValidatedSearchPathSettings {
+                extra_paths,
+                workspace_root,
+                stdlib,
+                site_packages,
+            },
         }
     }
 }
 
-/// A resolved module resolution order as per the [typing spec]
+#[derive(Debug, PartialEq, Eq, Clone)]
+struct ValidatedSearchPathSettings {
+    extra_paths: Vec<SearchPathRoot>,
+    workspace_root: SearchPathRoot,
+    stdlib: SearchPathRoot,
+    site_packages: Option<SearchPathRoot>,
+}
+
+impl ValidatedSearchPathSettings {
+    /// Implementation of the typing spec's [module resolution order].
+    ///
+    /// It's important that we collect these instead of iterating over them lazily,
+    /// so that we can be sure that it doesn't contain any duplicates.
+    /// At runtime, [`sys.path` never contains duplicate entries]
+    ///
+    /// [module resolution order]: https://typing.readthedocs.io/en/latest/spec/distributing.html#import-resolution-ordering
+    /// [`sys.path` never contains duplicate entries]: https://docs.python.org/3/library/site.html#module-site
+    fn static_search_paths(&self) -> OrderSet<&SearchPathRoot> {
+        let ValidatedSearchPathSettings {
+            extra_paths,
+            workspace_root,
+            stdlib,
+            site_packages,
+        } = self;
+        extra_paths
+            .iter()
+            .chain(std::iter::once(workspace_root))
+            .chain(std::iter::once(stdlib))
+            .chain(site_packages)
+            .collect()
+    }
+}
+
+/// Collect all dynamic search paths:
+/// search paths listed in `.pth` files in the `site-packages` directory
+/// due to editable installations of third-party packages.
+#[salsa::tracked(return_ref)]
+pub(crate) fn dynamic_module_resolution_paths(db: &dyn Db) -> OrderSet<SearchPathRoot> {
```

Comment:
I recommend that we call `db.report_untracked_read();` in the body. It informs salsa to only cache this query for the current revision. This ensures that the results are correct even when we don't track the `system.is_directory` call yet (at a small perf penalty that this query is executed every time we run a new check phase but I think that's neglectable).

Reply:
I think we should add a comment above the call to `db.report_untracked_read()` explaining what it does, why it's needed, and under what conditions we could remove it. (This should be a TODO comment if we intend to remove it at some point.)

Reply:
Added a big TODO